File: 1.HTTP-AND-CORS/code/go/layer_2_validation_via_etag_last_modifie_9.go
func getResource(w http.ResponseWriter, r *http.Request) {
    body := loadResource()
    sum := sha256.Sum256(body)
    etag := `"` + hex.EncodeToString(sum[:8]) + `"`  // fingerprint of the content

    if r.Header.Get("If-None-Match") == etag {  // client already has this exact version
        w.WriteHeader(http.StatusNotModified)   // 304, no body ,  payload saved
        return
    }
    w.Header().Set("ETag", etag)
    w.Header().Set("Cache-Control", "max-age=10")
    w.Write(body)                               // 200 + body
}

File: 1.HTTP-AND-CORS/code/python/layer_2_validation_via_etag_last_modifie_9.py
import hashlib
from fastapi import Request, Response

@app.get("/resource")
def get_resource(request: Request):
    body = load_resource()
    etag = '"' + hashlib.sha256(body).hexdigest()[:16] + '"'  # fingerprint

    if request.headers.get("if-none-match") == etag:   # already cached
        return Response(status_code=304)               # 304, no body ,  payload saved

    return Response(content=body, status_code=200,
                    headers={"ETag": etag, "Cache-Control": "max-age=10"})
